File: lab6/lab6_3.kt
interface iLogger {
  fun logMessage(msg: String)
}

abstract class Dispozitiv {
  protected var numeDispozitiv: String = ""

  fun getNume(): String { return numeDispozitiv }

  abstract fun esteConectat(): Boolean
}

class Imprimanta(nume: String) : iLogger, Dispozitiv() {
  private var conectat: Boolean = false

  init {
    numeDispozitiv = nume
  }

  fun conectare() {
    conectat = true
    logMessage("Imprimanta este acum conectata.")
  }

  fun deconectare() {
    conectat = false
    logMessage("Imprimanta a fost deconectata.")
  }

  override fun logMessage(msg: String) {
    println("[LOG $numeDispozitiv]: $msg")
  }

  override fun esteConectat(): Boolean {
    return conectat
  }
}

fun main() {
  val hp = Imprimanta("HP LaserJet")

  println("Nume dispozitiv: ${hp.getNume()}")

  if(!hp.esteConectat())
    println("Dispozitivul nu este conectat.")

  hp.conectare()

  if(hp.esteConectat())
    println("Dispozitivul este acum conectat.")

  hp.logMessage("Se printeaza un document test.")
  hp.deconectare()
}

File: lab6/lab6_3.cpp
#include <iostream>
#include <string>

using namespace std;

class iLogger {
public:
  virtual void logMessage(string msg) = 0;
};

class Dispozitiv {
public:
  string getNume() {
    return numeDispozitiv;
  }

  virtual bool esteConectat() = 0;

protected:
  string numeDispozitiv;
};

class Imprimanta : public iLogger, public Dispozitiv {
public:
  Imprimanta(string nume) {
    numeDispozitiv = nume;
    conectat = false;
  }

  void conectare() {
    conectat = true;
    logMessage("Imprimanta este acum conectata.");
  }

  void deconectare() {
    conectat = false;
    logMessage("Imprimanta a fost deconectata.");
  }

  void logMessage(string msg) override {
    cout << "[LOG " << numeDispozitiv << "]: " << msg << "\n";
  }

  bool esteConectat() override {
    return conectat;
  }

private:
  bool conectat;
};

int main() {
  Imprimanta hp("HP LaserJet");

  cout << "Nume dispozitiv: " << hp.getNume() << "\n";

  if(!hp.esteConectat())
    cout << "Dispozitivul nu este conectat.\n";

  hp.conectare();

  if(hp.esteConectat())
    cout << "Dispozitivul este acum conectat.\n";

  hp.logMessage("Se printeaza un document test.");
  hp.deconectare();

  return 0;
}